### ISSUE TITLE
TreeItem.getItemCount() and TreeItem.getItem(int) have linear complexity on Linux

### ISSUE THREAD

**Issue by basilevs** [eclipse-platform/eclipse.platform.swt#882]
The org.eclipse.swt.widgets.Tree is slow on Linux. In particular `TreeItem.getItemCount()` and `TreeItem.getItem(int)` have linear complexity (the time of execution is proportional to the number of children).

**To Reproduce**

```java
import org.eclipse.swt.SWT;
import org.eclipse.swt.layout.FillLayout;
import org.eclipse.swt.widgets.Shell;
import org.eclipse.swt.widgets.Tree;
import org.eclipse.swt.widgets.TreeItem;

public class TreeTest {

	private Shell shell = new Shell();
	{
		shell.setLayout(new FillLayout());
		shell.open();
	}


	public void getItemShouldHaveConstantTime() {
		for (int i = 0; i < 1000; i++) {
			measureGetItemTime(100); // warmup
		}
		double elapsed_10 = measureGetItemTime(10);
		double elapsed_100000 = measureGetItemTime(100000);
		double ratio = elapsed_100000 / elapsed_10;

		System.out.printf("Time for 10 elements: %f ns\nTime for 100000 elements: %f ns\nRatio: %f\n",elapsed_10, elapsed_100000, ratio);
	}

	private long measureGetItemTime(int count) {
		Tree tree = createTree();
		try {
			TreeItem subject = new TreeItem(tree, SWT.NONE);

			createChildren(subject, count);
			readAndDispatch();

			// warmup
			for (int i = 0; i < 1000; i++) {
				System.nanoTime();
				subject.getItem(count - 1);
			}

			long start = System.nanoTime();
			for (int i = 0; i < 10000; i++) {
				subject.getItem(count - 1);
			}
			long stop = System.nanoTime();
			long elapsed = stop-start;
			return elapsed;
		} finally {
			tree.dispose();
		}
	}

	private void createChildren(TreeItem subject, int count) {
		for (int i = 0; i < count; i++) {
			new TreeItem(subject, SWT.NONE);
		}
	}

	private Tree createTree() {
		Tree result = new Tree(shell, SWT.NONE);
		return result;
	}

	private void readAndDispatch() {
		for (int i = 0; i < 10; i++ ) {
			while(shell.getDisplay().readAndDispatch()) {}
		}
	}


}
```

This test wrapped in a product: 
[testporduct.zip](https://github.com/eclipse-platform/eclipse.platform.swt/files/13325209/testporduct.zip) (sources)
[slowGetItem.linux.gtk.x86_64.zip](https://github.com/eclipse-platform/eclipse.platform.swt/files/13325212/slowGetItem.linux.gtk.x86_64.zip) [slowGetItem.linux.gtk.aarch64.zip](https://github.com/eclipse-platform/eclipse.platform.swt/files/13325207/slowGetItem.linux.gtk.aarch64.zip)  (built product)

**Expected behavior**
Access time should not depend on the number of items.

**Environment:**
1. Select the platform(s) on which the behavior is seen:
- - [ ] All OS
- - [ ] Windows
- - [x] Linux
- - [ ] macOS 

2. Any GTK
4. Any JRE

**Version since**
Was there always?

**Workaround (or) Additional context**
Initially reported in https://github.com/eclipse-platform/eclipse.platform.ui/issues/649 Jface algorithms make navigation time proportional to a square of child count due to this problem.

Workaround: https://github.com/eclipse-platform/eclipse.platform.ui/pull/810

[GTK returns children count in O(N)](https://gitlab.gnome.org/GNOME/gtk/-/blob/gtk-3-24/gtk/gtktreestore.c#L753):

```c
static gint
gtk_tree_store_iter_n_children (GtkTreeModel *tree_model,
				GtkTreeIter  *iter)
{
  GNode *node;
  gint i = 0;

  g_return_val_if_fail (iter == NULL || iter->user_data != NULL, 0);

  if (iter == NULL)
    node = G_NODE (GTK_TREE_STORE (tree_model)->priv->root)->children;
  else
    node = G_NODE (iter->user_data)->children;

  while (node)
    {
      i++;
      node = node->next;
    }

  return i;
}
```

**Comment by basilevs** [comment]
I have not found a way to make `TreeItem.getItemCount()` without using this GTK API, but I have an idea of caching child count to help with JFace scenario, where child count does not change, but each child is accessed in sequence. Unfortunately, cache invalidation seems to quite complex with this approach.

**Comment by laeubi** [comment]
Just wondering, why do we need to query gtk at all? should not SWT/JFace be the one creating items and keep track of the count?

**Comment by basilevs** [comment]
> should not SWT/JFace be the one creating items and keep track of the count?

I'm under-qualified to answer, but one reason could be consistency with other platforms, where underlying framework is more efficient.

**Comment by SyntevoAlex** [comment]
Indeed GTK `gtk_tree_store_iter_n_children()` iterates a list of items to count children:
https://github.com/GNOME/gtk/blob/3.24.38/gtk/gtktreestore.c#L766-L771

Note that on Windows, SWT caches number of items:
https://github.com/eclipse-platform/eclipse.platform.swt/blob/eafdfbdfd7a4ee62a6bdb694be97f9e13abade09/bundles/org.eclipse.swt/Eclipse%20SWT/win32/org/eclipse/swt/widgets/Tree.java#L3340-L3348

**Comment by basilevs** [comment]
Naive caching helps with `childCount()`, but it turns out `getChilld(int)` is ~ (2N), and when the fix is applied it stays O(N). See my [PR comment](https://github.com/eclipse-platform/eclipse.platform.swt/pull/883#issuecomment-1806917114).

**Comment by laeubi** [comment]
@basilevs what I mean is that SWT itself is creating native items, so how could it be it does not knows (or could know) the number of items or the n-th item... so if the native implementation is to slow we probably should store an array of items on the java side that allows fast retrieval of` childCount()` **and** `getChilld(int)`

**Comment by basilevs** [comment]
@laeubi I've checked Windows implementation, and there TreeItems are not stored on Java side. The are created as needed using information from native code and then discarded. I guess this was the design goal for all platforms initially, but GTK binding has seemingly lost(?) the ability to discard TreeItem Java objects. Anyway, keeping another copy of the whole tree model seems to make VIRTUAL mode less valuable.

**Comment by laeubi** [comment]
We don't need a whole copy of the model:

1. Virtual only shows portions of the model that are visible
2. One only needs to store the pointers to the natives to get a fast count/getX by not iterate the whole model

**Comment by basilevs** [comment]
@laeubi as there  no way to detect visible object going out of scope, it is impossible to clean the cache, which means we would store all the objects that had ever been shown, which is not much better than whole model  (indeed in context of Jface, tis cache would have ALL items ever returned by content provider). 
2. Storing pointers is effectively same as storing reference to Java objects, given GTK implementation stores a cache of Java objects and has a way to transform pointers to references.

**Comment by basilevs** [comment]
We can avoid cache cleanup problem by only holding children of one item of interest, and repopulating it when interest shifts (Windows implementation does it for its caches).

**Comment by basilevs** [comment]
@laeubi I've implemented the cache and fixed the quadratic complexity of traversal, but the solution (in #883) stinks. I'm still thinking about either making TreeItem volatile or implementing your suggestion of full model mirror. Both are disruptive.

**Comment by basilevs** [comment]
I was wrong - efemeral Java TreeItems were never an option - strong references to them are held on all platforms.
The difference is : MacOS uses those strong references to keep track of parent-child relationship, while Linux and Windows create a convoluted system of identifiers to keep track of parents and children.

**Comment by laeubi** [comment]
@basilevs as slow Tree performance is a major concern I think it would be great to improve it even if it might be disruptive, if you can prepare things we can merge as soon as master is open and have a full cycle to test so don't worry as long as all tests pass :+1:

**Comment by basilevs** [comment]
I went with my dumb caching approach and it works for the scenario I'm interested in. It fails to actually make children tracking sensible, so I suggest not to close this issue until someone volunteers to rework SWT GTK to actually make `getItem()` constant time.

**Comment by iloveeclipse** [comment]
@SyntevoAlex : any chance you could look at #883 & remark there regarding the "complete" solution?

**Comment by SyntevoAlex** [comment]
I could have a look on wednesday, if I don't forget.

**Comment by basilevs** [comment]
The PR #883 is a failure. I've tested it on original testcase from https://github.com/eclipse-platform/eclipse.platform.ui/issues/649 and the polynomial growth is 1.8 which for practical purposes is not different from the original 2.0. I give up for now. Sorry for all the noise!

**Comment by basilevs** [comment]
An example of ID use case:

```
org.eclipse.swt.widgets.Tree.getFocusItem() {
	long [] path = new long [1];
	GTK.gtk_tree_view_get_cursor (handle, path, null);
	if (path [0] == 0) return null;
	TreeItem item = null;
	long iter = OS.g_malloc (GTK.GtkTreeIter_sizeof ());
	if (GTK.gtk_tree_model_get_iter (modelHandle, iter, path [0])) {
		int [] index = new int [1];
		GTK.gtk_tree_model_get (modelHandle, iter, ID_COLUMN, index, -1);
		if (index [0] != -1) item = items [index [0]]; //TODO should we be creating this item when index is -1?
	}
	OS.g_free (iter);
	GTK.gtk_tree_path_free (path [0]);
	return item;
}
```

Here we GTK return an iterator and we need a TreeItem it corresponds to. As iterator is not a key, Tree  has to use additional user data stored natively.

**Comment by basilevs** [comment]
I've [prototyped the strong reference tree representation](https://github.com/basilevs/eclipse.platform.swt/tree/issue_882_gtk_direct_access). Simple traversal of a wide tree is now linear.  The hotspot is no longer in navigation (indexes computation).  `TreeItem.setText()` is dramatically non-linear.
While `setText()` was not mentioned in the original description, with it `TreeViewer.reveal()` is still quadratic.
```
GTK.gtk_tree_store_set(long, long, int, byte[], int) line: not available [native method]	
TreeItem.setText(int, String) line: 1740	
TreeItem.setText(String) line: 1748	
Test_org_eclipse_swt_widgets_Tree.lambda$29(String, TreeItem) line: 1385	
0x00000028010aab20.accept(Object) line: not available	
Test_org_eclipse_swt_widgets_Tree.breadthFirstTraverse(TreeItem, Consumer<TreeItem>) line: 1262	
Test_org_eclipse_swt_widgets_Tree.lambda$28(TreeItem) line: 1385	
0x00000028010aa8f8.run() line: not available	
Test_org_eclipse_swt_widgets_Tree.measureNanos(Runnable) line: 1290	
Test_org_eclipse_swt_widgets_Tree.measureUpdateAllChildren(int) line: 1383	
0x00000028010aa4a0.apply(int) line: not available	
Test_org_eclipse_swt_widgets_Tree.assertLinear(String, IntFunction<Double>) line: 1315	
Test_org_eclipse_swt_widgets_Tree.lambda$26() line: 1376	
0x00000028010aa278.run() line: not available	
Test_org_eclipse_swt_widgets_Tree.testTreeRegularAndVirtual(Runnable) line: 1033	
Test_org_eclipse_swt_widgets_Tree.test_updateAllChildrenLinearGrowth() line: 1375	
```

```
java.lang.AssertionError: Update all children should be linear. But:
Time for 100 elements: 1341.017690 ns
Time for 100000 elements: 96560298.320000 ns
Ratio: 72005.238282
```

**Comment by basilevs** [comment]
Tree building and updatin of large trees is 20 times quicker on my prototype branch. This is achieved by removing as much non-constant GTK communication as possible and performing some of necessary communications lazily in cell update procedure.
Unfortunately, the polynomial approximation is still quardatic for the whole tree update (and correctness is questionable).
Simple sampling profiler no longer shows GTK operations as hotspots, I'm looking for other ways to isolate problematic cycles.

**Comment by basilevs** [comment]
Lazy Tree reveal is 340 times faster on the prototype branch for 100000 elements in the tree. The reval operation execution time is linear over the tree size. Other operations are also significantly quicker. 

Master branch:
```
213 ms to process 10000 items
218 ms to process 100000 items
The time to select scales as O(n^0.01) as the size of the model grows from 10000 to 100000
1249 ms to process 10000 items
93684 ms to process 100000 items
The time to reveal scales as O(n^1.87) as the size of the model grows from 10000 to 100000
```

The prototype branch:
```
(SWT:74185): Pango-WARNING **: 18:03:26.227: Invalid UTF-8 string passed to pango_layout_set_text()
224 ms to process 10000 items
238 ms to process 100000 items
The time to select scales as O(n^0.02) as the size of the model grows from 10000 to 100000

(SWT:74185): Pango-WARNING **: 18:03:28.701: Invalid UTF-8 string passed to pango_layout_set_text()

(SWT:74185): Pango-WARNING **: 18:03:28.701: Invalid UTF-8 string passed to pango_layout_set_text()
39 ms to process 10000 items
240 ms to process 100000 items
The time to reveal scales as O(n^0.78) as the size of the model grows from 10000 to 100000

```

### Main factors
- GTK.gtk_tree_model_get was used to associated TreeItem with GTK table row with a custom "ID" column and to proactively update GTK rows on setText(), setForeground(), setBackground(), setImage(), setFont()
- GTK.gtk_tree_model_get is linear over tree size
- JFace eagerly traverses elements of interest and their siblings on some operations (reveal is especially surprising in lazy trees, as one can't expect eager traversal of LazyContentProvider).

### Solution
Replace system of "ID" identifiers with path locators. Note, path locators take again linear time over tree size to locate an element, but they take no time at all to build, reducing eager  tree building to linear execution time (from quadratic).
Move all update procedures from setText(), setForeground(), setBackground(), setImage(), setFont() to `cellDataProc()`, simlarly to how VIRTUAL tree operates, but instead of data updates, do GTK updates lazily this prevents the need to do linear operations eagelry. Doing them only for visible items again reduces item update from a linear to constant execution time, resulting for linear (instead of quadratic) execution time for full tree refresh.

The prototype branch is https://github.com/eclipse-platform/eclipse.platform.swt/compare/master...basilevs:eclipse.platform.swt:issue_882_gtk_direct_access

It has bugs and is large and inconvenient to review. I plan to split it into smaller changesets and try to fix bugs and style. Before I proceed with this refinement however, I'd like to confirm with @SyntevoAlex if such a fundamental change of how SWT GTK operates have a chance to be accepted.

Thanks for your support and sorry for the noise.

**Comment by laeubi** [comment]
@basilevs that  sounds really promising, I think if you can extract some smaller changes that are easier to review I don't see a reason why such performance improvements can't be accepted.

Branch is currently closed, so its a good time to prepare such PRs already and then get them merged as early as possible.

**Comment by basilevs** [comment]
@laeubi @SyntevoAlex could you create a temporary branch for my changes, so that I could target my PRs without disrupting the master? In particular, I'd like to start with a set of failing tests.

**Comment by laeubi** [comment]
@basilevs what exactly do you want to archive with a temporary branch? If you like you can enable Github actions on your fork and then create any branches/prs there and they will build with these.

Another strategy is that you e.g. create `branchX` and open a PR for that, then you create `branchY` from `branchX` for some more experiments and so on.

**Comment by basilevs** [comment]
I want performance test results to be a part of the multi-stage review process.

I was not aware I could have my own actions. I was under impression they use environment secrets. Thanks for the hint.

I do not understand second strategy. Do you suggest I create branches in my fork and invite reviewers there to review parts of the work? This seems strange to me as once my repository is eventually removed, the important review history would be lost.

**Comment by laeubi** [comment]
Maybe you can explain why a normal branch/PR does not work? Are performance tests not executed there?

**Comment by basilevs** [comment]
@laeubi merging first PR with failing tests in the master branch will fail build for everyone.

**Comment by laeubi** [comment]
@basilevs but why do you need to merge, you can just

- add more commits to the branch
- branch from that branch if you want to do further experiements

**Comment by basilevs** [comment]
I need to merge the first stage of change so that second stage can be reviewed independently with clear source delta and  easily comparable test results.

As the process is so hard to convey, lets stop its discussion for now. I will have to disable failing tests on every stage of the work and it will be harder to impress reviewer without visible improvements to test results, but as long as everything is merged, it does not matter.

**Comment by laeubi** [comment]
> I need to merge the first stage of change so that second stage can be reviewed independently with clear source delta and easily comparable test results.

That's why you can simply use different commit for that in github you can then select that commit to only see that changes.

So if we create a temporary branch here or in your repo does not make any real difference. So maybe just try that approach and let us know when there is a problem and I'm sure we can help.

**Comment by SyntevoAlex** [comment]
Glancing at `gtk_tree_store_get_value()` I do not observe linear complexity.

It seems to derefence iterator and obtain the Tree's "row" in O(1).
https://github.com/GNOME/gtk/blob/3.24.38/gtk/gtktreestore.c#L658

Can you please clarify?

**Comment by basilevs** [comment]
> Glancing at `gtk_tree_store_get_value()` I do not observe linear complexity.
> It seems to derefence iterator and obtain the Tree's "row" in O(1). https://github.com/GNOME/gtk/blob/3.24.38/gtk/gtktreestore.c#L658
> Can you please clarify?

This is an experimental result - I've profiled tree traversal and found that ID setting is  a slowest part of traversal.
Once tree traversal was sped up by removing `getID(long, boolean)` method, I've noticed that tests are still slow due to `TreeItem.setText()`, `TreeItem.setForeground()`, etc. and traversal with such calls is one polynomial  degree slower than without. I did not find the root cause in C code, but assumed it was hidden in either renderers, JNI, or C allocator.

I've run the tests again and see that problem is not in getForeground() but in setForeground(), as I mostly focused on setters during my testing, I've mixed them up in the previous message.

Please see the test https://github.com/basilevs/eclipse.platform.swt/blob/3d7b9bedc5ad082e35b0db24835c533535e601ed/tests/org.eclipse.swt.tests/JUnit%20Tests/org/eclipse/swt/tests/junit/performance/Test_org_eclipse_swt_widgets_Tree.java

It's `setForeground()` and `secondTraverse()` methods demonstrate how execution time of a single `setForeground()` call depends on the size of dataset:

```
secondTraverse[Shape: STAR, virtual: false]
Execution time should grow as 1.200000 degree polynom. 
Time for 10000 elements: 4221486.000000 ns
Time for 100000 elements: 18886615.000000 ns
Ratio: 0.447393
Degree: 0.650689

setForeground[Shape: STAR, virtual: false]
Execution time should grow as 1.300000 degree polynom. 
Time for 10000 elements: 427697692.000000 ns
Time for 100000 elements: 33867289090.000000 ns
Ratio: 7.918511
Degree: 1.898644

```

Traversal of a prepopulated tree is linear, while  same traversal with a `setForeground()` call is quadratic. These test are done on a master branch (the branch I linked to is a prototype, so copy the test if you want to run it on master).

**Comment by SyntevoAlex** [comment]
Please use Java profiler on your test on original SWT code to identify what exactly is slow.
So far I have an impression that you're acting on false theory and therefore changing a lot more things than is actually needed.

**Comment by basilevs** [comment]
I will try to leave `getID()` in when extracting the first refactoring step from prototype and will present profiling results once that is done. The first step is to remove `gtk_tree_model_iter_nth_child` and similar calls from main path.
We can discuss the next step then with results at hand.

**Comment by basilevs** [comment]
On this screenshot of master branch profile (for JFace reveal case) `get_tree_store_set()` takes time comparable to the obvious offender `gtk_tree_model_iter_nth_child()` 
<img width="1217" alt="Screenshot 2023-11-26 at 19 06 05" src="https://github.com/eclipse-platform/eclipse.platform.swt/assets/650857/ccd38da8-acbb-417a-8a53-e6e92865e783">

**Comment by SyntevoAlex** [comment]
You previously said that `gtk_tree_model_get()` is culprit, and now say that `get_tree_store_set()` is culprit.

`...set` with many items could be slow indeed.

A few years ago, I also tried to optimize Tree/Table, and collected these profiling results for inserting into `Table`:
```
   - 99,09% main
	  - 89,62% gtk_list_store_insert_with_values
		 - 36,45% g_signal_emit
			- g_signal_emit_valist
			   - 19,13% signal_emit_unlocked_R.isra.0
				  - 12,32% g_closure_invoke
					 - 10,54% row_inserted_marshal
						+ 3,95% g_value_get_boxed
						+ 1,66% g_value_get_object
						+ 1,49% g_datalist_get_data
						  0,95% g_type_interface_peek
						  0,67% g_object_get_data
					   0,54% g_closure_ref
					1,87% g_hash_table_lookup
					0,80% g_mutex_unlock
					0,66% g_mutex_lock
			   + 5,63% g_value_unset
			   + 2,31% boxed_proxy_collect_value
			   + 1,91% g_value_init_from_instance
			   + 1,15% g_type_is_a
		 - 13,23% gtk_list_store_set_valist_internal
			- 9,50% gtk_list_store_real_set_value
				 2,79% g_sequence_set
			   - 2,17% _gtk_tree_data_list_value_to_node
					0,69% g_type_fundamental
				 1,57% g_slice_alloc
				 0,77% g_sequence_get
			  0,83% g_value_unset
			  0,63% g_type_value_table_peek
		 - 10,55% gtk_list_store_get_path.lto_priv.0
			- 3,68% gtk_tree_path_append_index
			   - 0,98% g_malloc
					0,63% malloc
			  3,63% node_get_pos
			+ 1,78% gtk_tree_path_new
		 - 8,74% g_sequence_insert_before
			- 3,22% node_insert_before
				 node_rotate
			- 1,80% node_new
				 g_slice_alloc
			  0,50% rotate_down
		 - 7,10% g_sequence_get_iter_at_pos
			  1,23% g_sequence_get_length
		 - 3,71% iter_is_valid
			  2,71% g_sequence_iter_get_sequence
		 - 2,42% gtk_tree_model_row_inserted
			- 1,64% g_type_check_instance_is_a
				 0,97% type_node_check_conformities_UorL
		 - 2,10% gtk_tree_path_free
			- 1,51% cfree@GLIBC_2.2.5
				 0,93% _int_free
		   1,65% g_sequence_get_length
		   1,43% g_slice_free1
	  - 9,45% g_object_unref
		 - gtk_list_store_finalize
			- 7,06% g_sequence_foreach_range
			   - _gtk_tree_data_list_free
				  + 2,99% g_type_is_a
					1,39% g_slice_free1
			+ 2,39% g_sequence_free
```

Here it can be seen that GTK wastes a lot of time to send `row-inserted` signal.
It also prepares tree path for it, see `gtk_list_store_get_path()`.

For a Tree, it would be `gtk_tree_store_get_path`:
https://github.com/GNOME/gtk/blob/3.24.38/gtk/gtktreestore.c#L580

Which counts tree's children to find child index. This is `O(N)`.

The stupid thing is that nobody is listening to this signal, so the entire `O(N)` effort is in vain.

**Comment by SyntevoAlex** [comment]
Frankly, I lost track of this discussion and the suggested changes.

When you blame something, I have a look at it, find no problems, and then you begin to blame something else.

I would like to have a short (and correct!) summary of what exactly is a problem and how are you going to fix it, in a couple sentences without referring to your tests and branch.

**Comment by SyntevoAlex** [comment]
Also, reading you between the lines, I understand that the code you're trying to fix is
1. Inserting too many items into Tree
2. Not using `SWT.VIRTUAL`
3. Inserts all items at once, instead of inserting children when parent is expanded

I would say that violating (2) and (3) are the bigger problem here.

**Comment by basilevs** [comment]
@SyntevoAlex errors like these are unavoidable - I never looked into SWT GTK until a couple of weeks ago.  Thank you for your help so far! The comments I'm leaving here are serving their purpose of getting a qualified help.
Providing an absolutely correct interpretation of the problem is a lot of work. Hopefully with your help we can find it, although, I would be happy as long as the performance is fixed without regressions.

The code to fix:
1. uses a necessary amount of items (I'm not sure what is "too many items" for a general purpose Tree control and assume that a 10^5 items is adequate, although up to a million items also is not outlandish)
1. uses  JFace TreeViewer with SWT.VIRTUAL and  Lazy Tree Content Provider (which implies SWT.VIRTUAL)
1. uses JFace (i.e. does not insert elements directly at all)
1. works fine on Windows and MacOS

**Comment by basilevs** [comment]
@laeubi I'm unable to target a Pull request to my own branch in main SWT repository. Could you clarify, In the workflow you suggest how can I create a PR like https://github.com/basilevs/eclipse.platform.swt/pull/1 but in a repository without commit rights?

**Comment by laeubi** [comment]
@basilevs lets say you have `branch1` (from `master`) then you create a new `branch2` but not from `master` but `branch1` ...

after that, create the PR here with target = `master` then you get a PR that has two commits and you can switch between these commits in the review tab.

**Comment by basilevs** [comment]
> @basilevs lets say you have `branch1` (from `master`) then you create a new `branch2` but not from `master` but `branch1` ...
> 
> after that, create the PR here with target = `master` then you get a PR that has two commits and you can switch between these commits in the review tab.

This process results in a PR that is created in my repository (fork). I can't expect @SyntevoAlex to work on review in inofficial repository, can I?

**Comment by laeubi** [comment]
As said, simply target the master `branch` here...

**Comment by laeubi** [comment]
> Ah, you suggest that reviewer is supposed to switch between commits! I see now.

Yes :-)

**Comment by basilevs** [comment]
@SyntevoAlex I've prepared a first stage of the fix.
Could you please review https://github.com/eclipse-platform/eclipse.platform.swt/pull/903/commits/9ff75d4de468301b9872a619f7304444a21427a2 ?

It eliminates an obvious hotspot and is trivial in nature, but should make further discussions easier by simplifying profiles.

**Comment by SyntevoAlex** [comment]
> This process results in a PR that is created in my repository (fork) because that's were target branch is located. I can't expect @SyntevoAlex to work on review in unofficial repository, can I?

I already lost track why you want PR in your own repository. To me it sounds reasonable that you could work in your branch and prepare small PRs from it once you've fixed something.

**Comment by basilevs** [comment]
Current status: https://github.com/eclipse-platform/eclipse.platform.swt/pull/918 is ready for review.

**Comment by basilevs** [comment]
Current status: abandoned due to lack of reviewers and priority shift.
